Make Explodes conditional and implement TS elite explosions

reaperrr opened the pull request:
And use that to replicate original `EXPLODES` elite ability.
Fixes a part of #13209.

Note: I don't know how exactly the damage was calculated in the original, so my choice was a bit arbitrary and might need some fine-tuning.

reaperrr committed:
Make Explodes conditional

reaperrr committed:
Add elite explosion to TS Sonic & Stealth Tanks

Matching original TS.

GraionDilach commented:
The EXPLODES veteran ability just applied `Explodes=yes` to the unit on the reached level and did nothing if the unit already had it enabled by then. Notsure if this takes that in account.

atlimit8 commented on OpenRA.Mods.Common/Traits/Explodes.cs:
```diff
@@ -74,6 +77,9 @@ void INotifyCreated.Created(Actor self)
 		void INotifyKilled.Killed(Actor self, AttackInfo e)
 		{
+			if (IsTraitDisabled)
+				return;
```
Neither `IsTraitDisabled`, nor `!self.IsInWorld` is a complex Boolean expression; perhaps it would be fine to just use `IsTraitDisabled || !self.IsInWorld `.

pchote commented on mods/ts/rules/gdi-vehicles.yaml:
```diff
@@ -278,6 +278,12 @@ SONIC:
+		RequiresCondition: !rank-elite
+	Explodes@ELITE:
+		RequiresCondition: rank-elite
+		Weapon: UnitExplodeElite
```
From @GraionDilach's comment, should these be normal `UnitExplode`?

reaperrr commented:
Updated.

rob-v commented on mods/ts/rules/gdi-vehicles.yaml:
```diff
@@ -278,6 +278,12 @@ SONIC:
+	Explodes@ELITE:
+		RequiresCondition: rank-elite
+		Weapon: UnitExplode
+		EmptyWeapon: UnitExplode
```
If Weapon and EmptyWeapon's default is UnitExplode, isn't it practice/better to not include it in .yaml?

reaperrr committed:
Change Explodes.Weapon default to null

FieldLoader.Require makes the default pointless, so it makes more sense to change it to null.

reaperrr commented:
What @rob-v pointed out made me realize some more issues with this trait #13444, so I'll adress that with this PR, too.

However, I'd like some feedback & decision on the details of the 2nd and 3rd point in #13444, I want to avoid having to redo the changes two or three times.

rob-v commented on mods/ts/rules/gdi-vehicles.yaml:
```diff
@@ -278,6 +278,11 @@ SONIC:
+		RequiresCondition: !rank-elite
+	Explodes@ELITE:
+		RequiresCondition: rank-elite
+		Weapon: UnitExplode
```
Could you please clarify for me why EmptyWeapon for !rank-elite has inherited value "UnitExplodeSmall", but EmptyWeapon for rank-elite has default value "UnitExplode"?
(so for tanks it is UnitExplodeSmall+UnitExplodeSmall, for Sonic UnitExplode+UnitExplode)

reaperrr replied:
To ensure that UnitExplode is used when rank-elite, even when the sonic tank was reloading when it got killed.

rob-v replied:
I meant not why you did it, but why/how it is done, that once it is inherited, once default value used, I must miss something.

rob-v reviewed (approved):
👍

atlimit8 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d040a21c2fccd6f4598c85de861a129623b1db23...72c82359159b2b3683583dc7fded37de86e1eea3)
